KEP-3545: Improved multi-numa alignment in Topology Manager (#3549)
kubernetes/enhancements: keps/sig-node/3545-improved-multi-numa-alignment/README.md

### Goals

* Enable TopologyManager to prefer sets of NUMA nodes with shorter NUMA distances for all TopologyManager policies
* Preserve all other properties of all policies

## Proposal

### Risks and Mitigations

| Risk                                             | Impact | Mitigation                                                                                                                                                                               |
|--------------------------------------------------|--------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Bugs in the implementation lead to kubelet crash | High   | Disable the policy option and restart the kubelet. The workload will run but resources allocation can be spread among NUMA nodes which are non-adjacent what can lead to higher latency. |

## Design Details

### Proposed Change
 
We propose to

> - add a new flag in Kubelet called `TopologyManagerPolicyOptions` in the kubelet config or command line argument called `topology-manager-policy-options` which allows the user to specify the Topology Manager policy option.

Reviewer: IIUC we want to replicate the policy options management we added in https://github.com/kubernetes/enhancements/issues/2625 , right?  Doing this for the cpumanager policy options required nontrivial discussion and the outcome was like the first of a kind for k8s. I for myself think this process is sustainable enough and makes sense, but we need to doublecheck we can replicate with topology manager soliciting feedback. I think this will emerge during the review of this KEP, let's make sure it happens though.

Reviewer 2: You are right, we wanted to replicate the policy options in CPUManager to be consistent how we handle new policies/options regarding resource management in Kubelet.

Reviewer 3: My hope is that it will be easier to digest this time around since there is already precedence. We should definitely pull in the same people that raised questions last time though. Can you /cc them here?

Author: /cc @derekwaynecarr

Reviewer: I'll look to the discussion happened on sig-arch and CC the relevant folks.

Reviewer: back in time we discussed with prod-readiness and in sig-arch with @johnbelamaric

Reviewer 3: Agreed. Since there actually is precedent now, I don't think it should be a problem to follow the same process, but we should get confirmation.

Reviewer 4: FWIW this seems reasonable to me but I will defer to those involved last time

- add a new topology manager option called `prefer-closest-numa-nodes`; if present, this option will enable further refinements of the existing `restricted` and `best-effort` policies, this option has no effect for `none` and `single-numa-node` policies.

Best-effort or restricted policy chooses resources that will fit into the smallest number of NUMA nodes.
This enhancement wants to change how the best hint is chosen from choosing narrower `NUMANodeAffinity` bitmask to choosing the one that is the narrowest and has the shortest average distance between NUMA nodes.

To summarize properties of `prefer-closest-numa-nodes` policy:

* Preserve all properties of `best-effort` and `restricted` policies
* Choose `NUMANodeAffinity` bitmask which is the narrowest and if the bitmask width is the same prefer the bitmask with smaller average distance.

When `prefer-closest-numa-nodes` policy is enabled, we need to retrieve information regarding distances between NUMA nodes.

> Right now Topology manager discovers Node layout using [CAdvisor API](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L40).

Reviewer: how do we want to add this dependency? IOW, does cadvisor require this KEP to be accepted before to accept this enhancement, or we can enhance cadvisor in parallel? If this KEP gets delayed for whatever reason, can we have the enhancement in place in cadvisor anyway?

Reviewer 2: My goal was to create PR to `cadvisor` in parallel but don't make it block this proposal if we won't be able to make it in before release.

Reviewer: This approach is good enough for me. If we can spell out in the KEP which timelines are looking at to remove the temporary code and depend on cadvisor (e.g.: consider this a beta or a GA blocker) that would be perfect.

Author: Graduation criteria was updated with requirement of cadvisor `MachineInfo` update for beta release

Reviewer 2: @fromanirh @knopt I have create a PR to add NUMA distances to `cadvisor` https://github.com/google/cadvisor/pull/3179

We would need to extend the `MachineInfo` struct with Distances field which would be an array of uint64 which will describe the distance between a given node and other nodes in the system.

> This information can be read from sysfs:

Reviewer: asking for better understanding myself (in general): is the kernel the best (only?) data source to assess NUMA distances?

Reviewer 2: So the Numa distance can be read from ACPI SLIT (system locality information table), which means that it is basically hardcoded inside firmware. Kernel exposes this information via `sysfs`. I don't know about any other method of retrieving this info.

Reviewer: Thanks, I've limited knowledge in this area and looking forward to learn more. From my perspective trusting the kernel here is perfectly fine.

```go 
const     distanceFile = "distance"

func (fs *realSysFs) GetDistance(nodePath string) (string, error) {
    distancePath := fmt.Sprintf("%s/%s", nodePath, distanceFile)
    distance, err := ioutil.ReadFile(distancePath)
    if err != nil {
        return "", err
    }
    return strings.TrimSpace(string(distance)), err
}

func getDistance(sysFs sysfs.SysFs, nodeDir string) ([]uint64, error) {
    rawDistance, err := sysFs.GetDistance(nodeDir)
    if err != nil {
        //Ignore if per-node info is not available.
        klog.Warningf("Found node without distance information, nodeDir: %s", nodeDir)
        return nil, nil
    }

    distances := []uint64{}
    for _, distance := range strings.Split(rawDistance, " ") {
        distanceInt, err := strconv.ParseUint(distance, 10, 64)
        if err != nil {
            return nil, fmt.Errorf("cannot convert %s to int", distance)
        }
        distances = append(distances, distanceInt)
    }

    return distances, nil
}
```

> But before this lands into `cadvisor`, this information would have to be discovered by `kubelet`.

Reviewer: how this will be handled? Will we have temporarily the same code above?

Reviewer 2: Yes.

### Implementation strategy

- Introduce new flag in Kubelet called `topology-manager-policy-options`, which when specified with `prefer-closest-numa-nodes` will modify the behavior of `best-effort` and `restricted` policy to pick NUMA nodes based on average distance between them.
- The `TopologyManagerPolicyOptions` flag is propagated to `ContainerManager` and later to `TopologyManager`.

> - When `TopologyManager` is being created it discovers distances between NUMA nodes and stores them inside `manager` struct. This is temporary until `distance` information will land into `cadvisor`.

Reviewer: How this temporary code path will be handled and how the phase off will be handled? are we looking for this temporary approach because there is uncertainity about the timeline of merging the new feature in cadvisor or there are other factors?
in general, I'd prefer to try very hard to avoid this temporary solution, because temporary solutions tend to stick around very long

Reviewer 3: There is precedence for this with the existence of the `GetNUMANode()` function we had in the original `TopologyManager` code while we waited for `cadvisor` to catch up with the info it supplied in its `MachineInfo` struct.

`GetNUMANode()`  was introduced here:
https://github.com/kubernetes/kubernetes/commit/ecc14fe661c22f5da967a7ff50cfb3aead60905b

And removed here:
https://github.com/kubernetes/kubernetes/commit/a047e8aa1b705bb7e5be881fb63cf90a218b60d0

@knopt, we should point at this in the KEP as an example of how this has been done in the past

Reviewer: I'm still not 100% happy with temporary approach, especially if enhancing cadvisor is compatible with the timelines we're looking at, but if there is consensus this approach is OK - and having a precedent surely helps - I'm *NOT* gonna block on this point here.

Author: Thanks for the comments, I've updated both implementation plan and graduation criteria. I've set cadvisor update to be blocker for alpha to beta graduation, and included the links to `GetNUMANode()` introduction and removal.

###### What happens if we reenable the feature if it was previously rolled back?

- No changes. The allocation of resources won't be changed for exiting containers. It will change for new containers only.

###### Are there any tests for feature enablement/disablement?

- There will be specific e2e test demonstrating that default behaviour is preserved when feature gate is disabled.

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

Kubelet may fail to start. The kubelet may crash.

###### What specific metrics should inform a rollback?

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No.

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

Inspect the kubelet configuration of the nodes: check feature gate and usage of the new option

###### How can someone using this feature know that it is working for their instance?

In order to verify this feature is working, one should:

> Pick a node with at least 4 NUMA nodes and with at least 3 different distance values between them.

Reviewer: this is hardly be possible in the current CI environment AFAIK. Are there plans to enable this?

Author: I'm not sufficiently familiar with the CI to answer this atm. We could add this as a graduation criteria for beta/GA, and start the discussion in sig-test to see if it's possible to cover this case in e2e tests.

Reviewer 3: That should be sufficient.

Reviewer 3: Can you point at where you added this as a graduation criteria?

I don't see it under the section:
```
**#### Graduation of Options to `Beta-quality` (non-hidden)**
```

Reviewer 2: I have added this as graduation criteria from Beta to GA [here](https://github.com/kubernetes/enhancements/pull/3549/files#diff-71a29ead6cac293d3c3eb9076377f00c3d6726b44ab2e66031e33d4da7dfb08aR316).

Reviewer 4: node e2e tests can choose which kind of machine to run against, e.g. https://github.com/kubernetes/test-infra/blob/master/jobs/e2e_node/image-config-serial-cpu-manager.yaml#L9 used by [pull-kubernetes-node-kubelet-serial-cpu-manager](https://github.com/kubernetes/test-infra/blob/f1fcaa2df0d8e2d01d3e419176b14e98874a49d9/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml#L459)

but I'm not sure if that will allow you to be as granular as you're describing

  - Verify no significant performance degradation

##### e2e tests

These cases will be added in the existing e2e tests:
  - Feature gate enable/disable tests
  - `prefer-closest-numa-nodes` policy option works as expected.

### Graduation Criteria

#### Alpha
- [X] Implement the new policy option.
- [X] Temporarily include NUMA distances discovery logic in kubelet code.
- [X] Add proper e2e node tests.

#### Alpha to Beta Graduation
- [X] Gather feedback from the consumer of the `prefer-closest-numa-nodes` policy option.
- [X] Remove NUMA distances discovery logic from kubelet in favor of updated cAdvisor.
- [X] No major bugs reported in the previous cycle.

#### Beta to G.A Graduation
- [X] Allowing time for feedback (1 year).
- [X] Risks have been addressed.

### Upgrade / Downgrade Strategy

We expect no impact. The new policiy options are opt-in.

### Version Skew Strategy

No changes needed

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: `TopologyManagerPolicyOptions`

Reviewer 3: I think we should go ahead and introduce `TopologyManagerPolicyAlphaOptions` now. I think it will be more confusing to retroactively add it later.

>   - Feature gate name: `TopologyManagerPolicyOptions`

Reviewer 2: have added all 3 feature gates here.

At present, the `TopologyManager` provides policies to align resources on either a single NUMA node or the minimum number of NUMA nodes (in cases where more than one NUMA node is required). However, the `TopologyManager` is not aware of NUMA distances and does not take them into account when making admission decisions.

This limitation surfaces in multi-socket, as well as single-socket multi NUMA systems (commonly present in AMD hardware), and can cause significant performance degradation in latency-critical execution and high-throughput applications if the `TopologyManager` decides to align resources on non-adjacent NUMA nodes.

### Goals

> * Enable `TopologyManager` to prefer sets of NUMA nodes with shorter NUMA distances for all TopologyManager policies

Reviewer 3: I would add a goal of:
* Introduce a new `kubelet` flag for passing options to a `TopologyManager` policy to modify its behaviour

Best-effort or restricted policy chooses resources that will fit into the smallest number of NUMA nodes.
This enhancement wants to change how the best hint is chosen from choosing narrower `NUMANodeAffinity` bitmask to choosing the one that is the narrowest and has the shortest average distance between NUMA nodes.

To summarize properties of `prefer-closest-numa-nodes` policy:

* Preserve all properties of `best-effort` and `restricted` policies
* Choose `NUMANodeAffinity` bitmask which is the narrowest and if the bitmask width is the same prefer the bitmask with smaller average distance.

When `prefer-closest-numa-nodes` policy is enabled, we need to retrieve information regarding distances between NUMA nodes.
Right now Topology manager discovers Node layout using [CAdvisor API](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L40).

We will need to extend the `MachineInfo` struct with a `Distances` field which will describe the distance between a given NUMA node and other NUMA nodes in the system.
This is already implemented in `cadvisor` by this [patch](https://github.com/google/cadvisor/pull/3179) but it is not yet present in any of the released versions. 

> Before new release of `cadvisor` includes this patch we will need to replicate this logic in `kubelet`.

Reviewer 3: ```suggestion
Until a new release of `cadvisor` includes this patch (and it gets vendored into the `kubelet`) we will need to replicate this logic in the `kubelet` code itself.
```

  - Verify no significant performance degradation

##### e2e tests

These cases will be added in the existing e2e tests:
  - Feature gate enable/disable tests
  - `prefer-closest-numa-nodes` policy option works as expected.

### Graduation Criteria

#### Alpha
- [X] Implement the new policy option.
- [X] Temporarily include NUMA distances discovery logic in kubelet code.
- [X] Add proper e2e node tests.

#### Alpha to Beta Graduation
- [X] Gather feedback from the consumer of the `prefer-closest-numa-nodes` policy option.
- [X] Remove NUMA distances discovery logic from kubelet in favor of updated cAdvisor.
- [X] No major bugs reported in the previous cycle.

#### Beta to G.A Graduation
- [X] Allowing time for feedback (1 year).
- [X] Risks have been addressed.
- [X] Presence of E2E test validating the described behavior on 4+ NUMA nodes system.

### Graduation Criteria of Options

In 1.26 we are releasing this feature to Alpha. We propose the following management of TopologyManager policy options graduation:

- `TopologyManagerPolicyAlphaOptions` is not enabled by default. Topology Manager alpha options (only one as of 1.26), are hidden by default

>  and only when `TopologyManagerPolicyAlphaOptions` feature gate is enabled user is able to use alpha options.

Reviewer 3: ```suggestion
 and only when both the `TopologyManagerPolicyOptions` and the `TopologyManagerPolicyAlphaOptions` feature gates are enabled will the user be able to use alpha options.
```

  `TopologyManagerPolicyAlphaOptions` will not exist if there are no options in alpha stage.

- `TopologyManagerPolicyBetaOptions` is not enabled by default. Topology Manager beta options (none as of 1.26), are hidden by default

>  and only when `TopologyManagerPolicyBetaOptions` feature gate is enabled user is able to use beta options.

Reviewer 3: ```suggestion
 and only when both the `TopologyManagerPolicyOptions`and the `TopologyManagerPolicyBetaOptions` feature gates are enabled will the user be able to use beta options.
```

`TopologyManagerPolicyBetaOptions` will not exist if there are no options in beta stage (true for 1.26 release).

- `TopologyManagerPolicyOptions` for enabling/disabling the entire feature. As this is an alpha feature, this feature gate would be disabled by default.

> Explicitly enabling `TopologyManagerPolicyOptions` feature gate provides us the ability to supply `TopologyManagerPolicyOptions` or `topology-manager-policy-options` flag in Kubelet.

Reviewer 3: Please move this to the top of the list.

When an option graduates, its visibility should be moved to be controlled by the corresponding feature-flag.
The introduction of these feature gates gives us the ability to move the option to beta and later stable without implying that all available options are stable.

> 

Reviewer 3: I would add a reference to how there is existing precedence for doing the same in the cpumanager here.

Reviewer 2: added

#### Graduation of Options from `Beta-quality` to `G.A-quality`
- [X] Allowing time for feedback (1 year) on the policy option.
- [X] Risks have been addressed.

### Upgrade / Downgrade Strategy

We expect no impact. The new policiy options are opt-in.

### Version Skew Strategy

No changes needed

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate names:
    - `TopologyManagerPolicyAlphaOptions`
    - `TopologyManagerPolicyBetaOptions`
    - `TopologyManagerPolicyOptions`
  - Components depending on the feature gate: kubelet
- [x] Change the kubelet configuration to set a `TopologyManager` policy to `restricted` or `best-effort` and a `TopologyManagerPolicyOptions` to `prefer-closest-numa-nodes`
  - Will enabling / disabling the feature require downtime of the control
    plane? 
    No.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node?
    Yes, a kubelet restart is required.

###### Does enabling the feature change any default behavior?

> - Yes, it makes the behaviour of the TopologyManager restricted and best-effort policies to choose NUMA nodes based on distance.

Reviewer 3: I'd argue the answer here is actually 'no'. It does not change any *default* behaviour -- the behaviour only changes when the flag is provided.

Reviewer 2: Yeah, you're right. Changed.

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> We propose an enhancement to TopologyManager that enables it to favor sets of NUMA nodes with shorter distance between the nodes, when making admission decision. The proposed enhancement is only applicable when comparing sets of NUMA nodes equal in size.

Reviewer 3: ```suggestion
We propose an enhancement to the `TopologyManager` that allows it to favor sets of NUMA nodes with shorter distance between nodes when making admission decisions. The proposed enhancement is only applicable when comparing sets of NUMA nodes that are equal in size.
```

## Motivation

> To support latency-critical execution and high-throughput applications (including those running on Kubernetes), NUMA locality(topology) of resources assigned to the workloads is crucial. THe best NUMA locality can be achieved by minimizing number of NUMA nodes required to execute the workload, and minimizing the NUMA distance between the nodes. NUMA distance is a metric which exposes relative physical distance between NUMA cells which relates to latency that is introduced when a CPU needs to access resources from different NUMA nodes. The distance is always within range 10-254 and for a local access its value is 10.

Reviewer 3: ```suggestion
To support latency-critical execution and high-throughput applications (including those running on Kubernetes), NUMA locality (topology) of resources assigned to the workloads is crucial. The best NUMA locality can be achieved by minimizing the number of NUMA nodes required to execute the workload, and minimizing the NUMA distance between those nodes. NUMA distance is a metric which exposes relative physical distance between NUMA cells which relates to latency that is introduced when a CPU needs to access resources from different NUMA nodes. The distance is always within range 10-254 and for a local access its value is 10.
```

> As of now, TopologyManager provides policies to either provide resources from only single numa node or the smallest number of NUMA nodes. TopologyManager is not aware of the NUMA distances and does not take it into consideration during admission decisions. 

Reviewer 3: ```suggestion
At present, the `TopologyManager` provides policies to align resources on either a single NUMA node or the minimum number of NUMA nodes (in cases where more than one NUMA node is required). However, the `TopologyManager is not aware of NUMA distances and does not take them into account when making admission decisions. 
```

> This limitation surfaces in multi socket, as well as single socket multi NUMA systems (commonly present in AMD hardware), and might cause a significant performance degradation for latency-critical execution and high-throughput applications, in a case when the Topology Manager schedules a pod on non-adjacent numa nodes.

Reviewer 3: ```suggestion
This limitation surfaces in multi-socket, as well as single-socket multi NUMA systems (commonly present in AMD hardware), and can cause significant performance degradation in latency-critical execution and high-throughput applications if the `TopologyManager` decides to align resources on non-adjacent NUMA nodes.
```

### Goals

> * Enable TopologyManager to prefer sets of NUMA nodes with shorter NUMA distances for all TopologyManager policies

Reviewer 3: ```suggestion
* Enable `TopologyManager` to prefer sets of NUMA nodes with shorter NUMA distances for all TopologyManager policies
```

- add a new topology manager option called `prefer-closest-numa-nodes`; if present, this option will enable further refinements of the existing `restricted` and `best-effort` policies, this option has no effect for `none` and `single-numa-node` policies.

Best-effort or restricted policy chooses resources that will fit into the smallest number of NUMA nodes.
This enhancement wants to change how the best hint is chosen from choosing narrower `NUMANodeAffinity` bitmask to choosing the one that is the narrowest and has the shortest average distance between NUMA nodes.

To summarize properties of `prefer-closest-numa-nodes` policy:

* Preserve all properties of `best-effort` and `restricted` policies
* Choose `NUMANodeAffinity` bitmask which is the narrowest and if the bitmask width is the same prefer the bitmask with smaller average distance.

When `prefer-closest-numa-nodes` policy is enabled, we need to retrieve information regarding distances between NUMA nodes.
Right now Topology manager discovers Node layout using [CAdvisor API](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L40).

> We would need to extend the `MachineInfo` struct with Distances field which would be an array of uint64 which will describe the distance between a given node and other nodes in the system.

Reviewer 3: ```suggestion
We will need to extend the `MachineInfo` struct with a `Distances` field which will describe the distance between a given NUMA node and other NUMA nodes in the system.
```

    distancePath := fmt.Sprintf("%s/%s", nodePath, distanceFile)
    distance, err := ioutil.ReadFile(distancePath)
    if err != nil {
        return "", err
    }
    return strings.TrimSpace(string(distance)), err
}

func getDistance(sysFs sysfs.SysFs, nodeDir string) ([]uint64, error) {
    rawDistance, err := sysFs.GetDistance(nodeDir)
    if err != nil {
        //Ignore if per-node info is not available.
        klog.Warningf("Found node without distance information, nodeDir: %s", nodeDir)
        return nil, nil
    }

    distances := []uint64{}
    for _, distance := range strings.Split(rawDistance, " ") {
        distanceInt, err := strconv.ParseUint(distance, 10, 64)
        if err != nil {
            return nil, fmt.Errorf("cannot convert %s to int", distance)
        }
        distances = append(distances, distanceInt)
    }

    return distances, nil
}
```

### Implementation strategy

- Introduce new flag in Kubelet called `topology-manager-policy-options`, which when specified with `prefer-closest-numa-nodes` will modify the behavior of `best-effort` and `restricted` policy to pick NUMA nodes based on average distance between them.

> - The `TopologyManagerPolicyAlphaOptions` flag is propagated to `ContainerManager` and later to `TopologyManager`.

Reviewer 3: Flag or featuregate?

Reviewer 2: it should be flag 😄

- Enable `TopologyManager` NUMA distances discovery:
  - Temporarily add distances discovery logic into `kubelet` (similarly to [the introduction](https://github.com/kubernetes/kubernetes/commit/ecc14fe661c22f5da967a7ff50cfb3aead60905b) of `GetNUMANode()`).
  - Once `cadvisor` exposes the NUMA distance information through `MachineInfo`, remove the logic out of `kubelet` (similarly to [the removal](https://github.com/kubernetes/kubernetes/commit/a047e8aa1b705bb7e5be881fb63cf90a218b60d0) of `GetNUMANode()`).
  - The removal of the logic `kubelet` is [an Alpha to Beta graduation criteria](#alpha-to-beta-graduation).

> - When `TopologyManager` is being created it discovers distances between NUMA nodes and stores them inside `manager` struct. This is temporary until `distance` information will land into `cadvisor`.

Reviewer 3: ```suggestion
- When `TopologyManager` is being created it discovers distances between NUMA nodes and stores them inside `manager` struct. This is temporary until `distance` information lands in `cadvisor`.
```

```go 

// NUMADistance is a matrix representing distances between NUMA nodes
type NUMADistance [][]uint64

func (n NUMADistance) CalculateAvgDistanceFor(bm bitmask.BitMask) int {
   // implementation
   return avgDistance
}

type PolicyOpts struct {
    PreferClosestNuma bool
    Distances NUMADistance
}

func compareHints(bestNonPreferredAffinityCount int, current *TopologyHint, candidate *TopologyHint, policyOpts *PolicyOpts) *TopologyHint {
    /*
    ...
    */

	if current.Preferred && candidate.Preferred {
        if candidate.NUMANodeAffinity.IsNarrowerThan(current.NUMANodeAffinity) {
            return candidate
        } 
        if policyOpts.PreferClosestNuma && candidate.NUMANodeAffinity.IsEqual(current.NUMANodeAffinity) {
            candidateDistance := policyOpts.Distances.CalculateAvgDistanceFor(candidate)
            currentDistance := policyOpts.Distances.CalculateAvgDistanceFor(current)
            // candidate avg distance is lower
            if candidateDistance < currentDistance {
                return candidate
            } 

            return current 
        }

> 	}

Reviewer 3: Please fix the indentation here.

    /*
    ...
    */
}

```

### Calculating average distance

Let's consider following distance table:

|  node/node | node0 | node1 | node2 | node3 |
| ---------- | ----- | ----- | ----- | ----- |
| node0      | 10    | 11    | 12    | 12    |
| node1      | 11    | 10    | 12    | 12    |
| node2      | 12    | 12    | 10    | 11    |
| node3      | 12    | 12    | 11    | 10    |

> If resources cannot be fitted into one NUMA node the new policy option will prefer hints with bitmasks that have lower average distance between NUMA nodes. Such bitmasks:

Reviewer 3: ```suggestion
If resources cannot fit onto one NUMA node, the new policy option will prefer hints with bitmasks that have lower average distance between NUMA nodes. Such bitmasks include:
```

* 1100 -> (10 + 11 + 11 + 10) /4 = 10.5
* 1010 -> (10 + 12 + 10 + 12) /4 = 11
* 1001 -> (10 + 12 + 10 + 12) /4 = 11
* 0011 -> (10 + 11 + 10 + 11 ) /4 = 10.5
* 0110 -> (10 + 12 + 10 + 12 ) /4 = 11

> So the bitmasks 1100 and 0011 has the lowest average distance between NUMA nodes.

Reviewer 3: ```suggestion
So the bitmasks 1100 and 0011 have the lowest average distance between NUMA nodes.
```

> If we consider system with 8 NUMA nodes:

Reviewer 3: ```suggestion
If we consider a system with 8 NUMA nodes:
```

|  node/node | node0 | node1 | node2 | node3 | node4 | node5 | node6 | node7 |
| ---------- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| node0      | 10    | 11    | 12    | 12    | 30    | 30    | 30    | 30    |
| node1      | 11    | 10    | 12    | 12    | 30    | 30    | 30    | 30    |
| node2      | 12    | 12    | 10    | 11    | 30    | 30    | 30    | 30    |
| node3      | 12    | 12    | 11    | 10    | 30    | 30    | 30    | 30    |
| node4      | 30    | 30    | 30    | 30    | 10    | 11    | 12    | 12    |
| node5      | 30    | 30    | 30    | 30    | 11    | 10    | 12    | 12    |
| node6      | 30    | 30    | 30    | 30    | 12    | 12    | 10    | 11    |
| node7      | 30    | 30    | 30    | 30    | 12    | 12    | 11    | 10    |

> And following bitmasks:

Reviewer 3: ```suggestion
And the following bitmasks:
```

* 10001000 -> (10 + 30 + 10 +30) /4 = 20
* 11100000 -> (10 + 11 + 12 + 10 + 11 + 12 + 12 + 12 + 10) / 6 = 16.7

> In second case even though the average distance is lower but the bitmask width is bigger and that is why the first case should be preffered.

Reviewer 3: ```suggestion
In the second case, even though the average distance is lower, the number of NUMA nodes is larger (meaning that the first case should always be preferred).
```

Best-effort or restricted policy chooses resources that will fit into the smallest number of NUMA nodes.
This enhancement wants to change how the best hint is chosen from choosing narrower `NUMANodeAffinity` bitmask to choosing the one that is the narrowest and has the shortest average distance between NUMA nodes.

To summarize properties of `prefer-closest-numa-nodes` policy:

* Preserve all properties of `best-effort` and `restricted` policies
* Choose `NUMANodeAffinity` bitmask which is the narrowest and if the bitmask width is the same prefer the bitmask with smaller average distance.

When `prefer-closest-numa-nodes` policy is enabled, we need to retrieve information regarding distances between NUMA nodes.
Right now Topology manager discovers Node layout using [CAdvisor API](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L40).

We will need to extend the `MachineInfo` struct with a `Distances` field which will describe the distance between a given NUMA node and other NUMA nodes in the system.

> This information can be read from sysfs:

Reviewer 5: Now that https://github.com/google/cadvisor/pull/3179 has been merged, can we remove references to the workaround (of obtaining distance information from `sysfs`)? Does the release cadence of cadvisor allow us to pull in the latest changes in Kubernetes 1.26 or we anticipate some lag and are intentionally keeping the workaround still here?

Reviewer 2: This is merged and available on `master` branch but not in any of official releases. I am not familiar with how `cadvisor` release schedule and I can't guarantee that this change will be released before Kubernetes `v1.26` that is why I have kept this workaround here.

Reviewer 5: Sure, thanks for clarifying. We can keep an eye on cadvisor release and adapt accordingly during the development phase depending on whether or not it makes its way into Kubernetes 1.26.

            return nil, fmt.Errorf("cannot convert %s to int", distance)
        }
        distances = append(distances, distanceInt)
    }

    return distances, nil
}
```

### Implementation strategy

- Introduce new flag in Kubelet called `topology-manager-policy-options`, which when specified with `prefer-closest-numa-nodes` will modify the behavior of `best-effort` and `restricted` policy to pick NUMA nodes based on average distance between them.
- The `TopologyManagerPolicyOptions` flag is propagated to `ContainerManager` and later to `TopologyManager`.
- Enable `TopologyManager` NUMA distances discovery:
  - Temporarily add distances discovery logic into `kubelet` (similarly to [the introduction](https://github.com/kubernetes/kubernetes/commit/ecc14fe661c22f5da967a7ff50cfb3aead60905b) of `GetNUMANode()`).
  - Once `cadvisor` exposes the NUMA distance information through `MachineInfo`, remove the logic out of `kubelet` (similarly to [the removal](https://github.com/kubernetes/kubernetes/commit/a047e8aa1b705bb7e5be881fb63cf90a218b60d0) of `GetNUMANode()`).
  - The removal of the logic `kubelet` is [an Alpha to Beta graduation criteria](#alpha-to-beta-graduation).

> - When `TopologyManager` is being created it discovers distances between NUMA nodes and stores them inside `manager` struct. This is temporary until `distance` information lands in `cadvisor`.

Reviewer 5: This might need adjustment if we decide to obtain distance information from cadvisor (given that the [PR](https://github.com/google/cadvisor/pull/3179) is already merged).

Reviewer 2: I can put link to merged PR and state that if `cadvisor` won't release this change before `v1.26` we will just replicate how distance is calculated?

Reviewer 2: changed.

### Proposed Change
 
We propose to

> - add a new flag in Kubelet called `TopologyManagerPolicyAlphaOptions` in the kubelet config or command line argument called `topology-manager-policy-options` which allows the user to specify the Topology Manager policy option.

Reviewer 5: I think the flag should still be called `TopologyManagerPolicyOptions` (corresponding to `topology-manager-policy-options` flag in kubelet). 

May be we are confusing flags with feature gates? Kubelet flags are different from Feature gates (the same name here is probably not helping!). In this case we we would need a kubelet flag (`TopologyManagerPolicyOptions` ) and three feature gates:

1.  `TopologyManagerPolicyOptions` for enabling/disabling the entire feature. As this is a brand new feature introduced as alpha, the feature gate would be disabled by default. Explicitly enabling `TopologyManagerPolicyOptions` feature gate provides us the ability to supply `topology-manager-policy-options` flag in Kubelet. 
2.  `TopologyManagerPolicyAlphaOptions` to enable/disable the visibility of options at alpha level options. These options would be hidden by default.
3.  `TopologyManagerPolicyBetaOptions` to enable/disable the visibility beta level. These options would be visible by default.

IOW, to use `topology-manager-policy-options=prefer-closest-numa-nodes`  in Kubernetes 1.26 we would have to specify both `TopologyManagerPolicyOptions` and `TopologyManagerPolicyAlphaOptions`. Once we graduate the overall feature to Beta `TopologyManagerPolicyOptions` is enabled by default and `prefer-closest-numa-nodes` moves from an alpha option to a Beta option (controlled by feature gate`TopologyManagerPolicyBetaOptions` feature gate) and is therefore visible by default without the need to providing a feature gate. In case of Beta options, `TopologyManagerPolicyBetaOptions` can be used to disable these policy options.

Hope this helps and provides more clarity.

In my comment [here](https://github.com/kubernetes/enhancements/pull/3549#discussion_r985636985), I was suggesting that we could get away with a single feature gate as this is an alpha feature and we only have a single option but it probably makes sense to capture the feature gates for the options to indicate maturity level of options right from the beginning to document and set expectations for options proposed in the future.

Reviewer 3: I agree with everything stated here. It's not stated explicitly here, but I want to emphasize that we should actually add the feature gate for `TopologyManagerPolicyBetaOptions` **now** (even though there is nothing to gate this with option yet). That way there is less code churn once we do graduate this feature to beta in a future release.

Reviewer 2: Fixed.

- Introduce new flag in Kubelet called `topology-manager-policy-options`, which when specified with `prefer-closest-numa-nodes` will modify the behavior of `best-effort` and `restricted` policy to pick NUMA nodes based on average distance between them.
- The `TopologyManagerPolicyAlphaOptions` flag is propagated to `ContainerManager` and later to `TopologyManager`.
- Enable `TopologyManager` NUMA distances discovery:
  - Temporarily add distances discovery logic into `kubelet` (similarly to [the introduction](https://github.com/kubernetes/kubernetes/commit/ecc14fe661c22f5da967a7ff50cfb3aead60905b) of `GetNUMANode()`).
  - Once `cadvisor` exposes the NUMA distance information through `MachineInfo`, remove the logic out of `kubelet` (similarly to [the removal](https://github.com/kubernetes/kubernetes/commit/a047e8aa1b705bb7e5be881fb63cf90a218b60d0) of `GetNUMANode()`).
  - The removal of the logic `kubelet` is [an Alpha to Beta graduation criteria](#alpha-to-beta-graduation).
- When `TopologyManager` is being created it discovers distances between NUMA nodes and stores them inside `manager` struct. This is temporary until `distance` information will land into `cadvisor`.

> - Pass `TopologyManagerPolicyAlphaOptions` to best-effort and restricted policy. When this is specified best-hint is picked based on average distance between NUMA nodes. This would require modification to `compareHints` function to change how the best hint is calculated:

Reviewer 5: This should be called `TopologyManagerPolicyOptions` (corresponding to topology-manager-policy-options flag in kubelet).

Reviewer 2: Fixed.

- [X] Allowing time for feedback (1 year).
- [X] Risks have been addressed.

### Graduation Criteria of Options

In 1.26 we are releasing this feature to Alpha. We propose the following management of TopologyManager policy options graduation:
- `TopologyManagerPolicyAlphaOptions` is not enabled by default. Topology Manager alpha options (only one as of 1.26), are hidden by default
 and only when `TopologyManagerPolicyAlphaOptions` feature gate is enabled user is able to use alpha options.
  `TopologyManagerPolicyAlphaOptions` will not exist if there are no options in alpha stage.
- `TopologyManagerPolicyBetaOptions` is not enabled by default. Topology Manager beta options (none as of 1.26), are hidden by default
 and only when `TopologyManagerPolicyBetaOptions` feature gate is enabled user is able to use beta options.
`TopologyManagerPolicyBetaOptions` will not exist if there are no options in beta stage (true for 1.26 release).

> - `TopologyManagerPolicyOptions` is enabled by default. Topology Manager stable options (none as of 1.26), are visible by default

Reviewer 5: Since this is an alpha feature, `TopologyManagerPolicyOptions` should be disabled by default.

Reviewer 3: Yes. This needs to be updated.

Reviewer 2: Fixed.

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: `TopologyManagerPolicyAlphaOptions`

Reviewer 5: We should add all the feature gates here:  `TopologyManagerPolicyOptions`, `TopologyManagerPolicyAlphaOptions` and `TopologyManagerPolicyBetaOptions`

Reviewer 2: Fixed.

  - Components depending on the feature gate: kubelet

> - [x] Change the kubelet configuration to set a `TopologyManager` policy to `restricted` or `best-effort` and a `TopologyManagerPolicyAlphaOptions` to `prefer-closest-numa-nodes`

Reviewer 5: This should be called TopologyManagerPolicyOptions.

Reviewer 2: Fixed.
